src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/ProjectFileClassifier.cs

```diff
@@ -0,0 +1,113 @@
+            }
+        }
+
+        /// <summary>
+        /// Gets and sets the paths found in the <c>NuGetPackageFolders</c> property value for this project.
+        /// Project files under any of these folders are considered non-modifiable.
+        /// </summary>
+        /// <remarks>
+        /// Example value: <c>"C:\Users\myusername\.nuget\;D:\LocalNuGetCache\"</c>
+        /// </remarks>
+        public string NuGetPackageFolders
+        {
+            set
+            {
+                Requires.NotNull(value, nameof(value));
+
+                if (!string.Equals(_nuGetPackageFoldersString, value, StringComparisons.Paths))
+                {
+                    _nuGetPackageFoldersString = value;
+                    _nuGetPackageFolders = value.Split(Delimiter.Semicolon, StringSplitOptions.RemoveEmptyEntries);
+                }
+            }
+        }
+
+        public ProjectFileClassifier()
+        {
+            _windows = Environment.GetFolderPath(Environment.SpecialFolder.Windows);
+            _programFiles86 = Environment.GetFolderPath(Environment.SpecialFolder.ProgramFilesX86);
+
+            // In a 32-bit process, SpecialFolder.ProgramFiles returns the 32-bit path.
+            // The 64-bit path is available via an environment variable however.
+            _programFiles64 = Environment.GetEnvironmentVariable("ProgramW6432");
+
+            _vsInstallationDirectory = GetVSInstallationDirectory();
+
+            EnsureTrailingSlash(ref _windows);
+            EnsureTrailingSlash(ref _programFiles86);
+            EnsureTrailingSlash(ref _programFiles64);
+            EnsureTrailingSlash(ref _vsInstallationDirectory);
+
+            return;
+
+            static string? GetVSInstallationDirectory()
+            {
+                string? dir = Environment.GetEnvironmentVariable("VSAPPIDDIR");
+
+                // The path provided is not the installation root, but rather the location of devenv.exe.
+                // __VSSPROPID.VSSPROPID_InstallDirectory has the same value.
+                // Failing a better way to obtain the installation root, remove that suffix.
+                // Obviously this is brittle against changes to the relative path of devenv.exe, however that seems
+                // unlikely and should be easy to work around if ever needed.
+                const string DevEnvExeRelativePath = "Common7\\IDE\\";
+
+                if (dir?.EndsWith(DevEnvExeRelativePath, StringComparisons.Paths) == true)
+                {
+                    dir = dir.Substring(0, dir.Length - DevEnvExeRelativePath.Length);
+                }
+
+                return dir;
+            }
+        }
+
+        private static void EnsureTrailingSlash([AllowNull] ref string path)
+        {
+            if (path is not null)
+            {
+                path = PathHelper.EnsureTrailingSlash(path);
+            }
+        }
+
+        public bool IsNonModifiable(string importPath)
```

Comment:
>IsNonModifiable [](http://example.com/codeflow?start=20&length=15)

What do you think about inverting this to be IsModifiable? It might make for easier reading where it is used, e.g.

if (obj.IsModifiable())...
if (!obj.IsModifiable())...

vs

if (!obj.IsNonModifiable())...
if (obj.IsNonModifiable())...

Reply:
I agree it's good to avoid a double negative like this, but as this is a heuristic there are three classification states:

1. Non-modifiable
2. Modifiable
3. 🤷‍♀️

Negating 1 doesn't give you 2.

Reply:
Accursed tri-state booleans! 😡

Reply:
Thinking about it... What about using a different word? Like, `IsImmutable`? That would not making it seem like a boolean situation.

Reply:
After discussion in last week's CPS meeting we concluded we need two tests:

1. Should the user be able to edit this file in the IDE?
2. Do we reasonably expect this file will not be modified once it has been created on disk?

My first sense is that these would be called:

1. `IsUserEditable`
2. `IsFileModifiable`

These aren't perfect names and so the API docs will have to fill in the gaps a bit. I feel "immutable" is too strong of a guarantee for these tests so I'm reluctant to use that term.